Content.Server/Revenant/EntitySystems/RevenantSystem.Abilities.cs

```diff
@@ -0,0 +1,302 @@
+            component.EssenceRegenCap += component.MaxEssenceUpgradeAmount;
+        }
+
+        if (TryComp<MobStateComponent>(args.Target, out var mobstate))
+        {
+            var damage = _mobState.GetEarliestDeadState(mobstate, 0)?.threshold;
+            if (damage != null)
+            {
+                DamageSpecifier dspec = new();
+                dspec.DamageDict.Add("Cellular", damage.Value);
+                _damage.TryChangeDamage(args.Target, dspec, true);
+            }
+        }
+    }
+
+    private void OnHarvestCancelled(EntityUid uid, RevenantComponent component, HarvestDoAfterCancelled args)
+    {
+        if (TryComp<AppearanceComponent>(uid, out var app))
+            app.SetData(RevenantVisuals.Harvesting, false);
+    }
+
+    private void OnDefileAction(EntityUid uid, RevenantComponent component, RevenantDefileActionEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        if (!CanUseAbility(uid, component, component.DefileCost, component.DefileDebuffs))
+            return;
+
+        args.Handled = true;
+
+        var lookup = _lookup.GetEntitiesInRange(uid, component.DefileRadius, LookupFlags.Approximate | LookupFlags.Anchored);
+
+        for (var i = 0; i < component.DefileTilePryAmount; i++)
+        {
+            //get random coordinates in the radius (technically a square but shut up)
+            var coords = new EntityCoordinates(uid,
+                (_random.NextFloat(-component.DefileRadius, component.DefileRadius), _random.NextFloat(-component.DefileRadius, component.DefileRadius)));
+
+            var gridID = coords.GetGridUid(EntityManager);
+            if (_mapManager.TryGetGrid(gridID, out var map))
+                map.GetTileRef(coords).PryTile(_mapManager, entityManager: EntityManager, robustRandom: _random);
+        }
+
+        foreach (var ent in lookup)
+        {
+            //break windows
+            if (HasComp<TagComponent>(ent) && _tag.HasAnyTag(ent, "Window"))
+            {
+                //hardcoded damage specifiers til i die.
+                var dspec = new DamageSpecifier();
+                dspec.DamageDict.Add("Structural", 15);
+                _damage.TryChangeDamage(ent, dspec);
+            }
+
+            //randomly opens some lockers and such.
+            if (_random.Prob(component.DefileEffectChance) && HasComp<EntityStorageComponent>(ent))
+                _entityStorage.OpenStorage(ent);
+
+            //chucks shit
+            if (_random.Prob(component.DefileEffectChance) && HasComp<ItemComponent>(ent) &&
+                TryComp<PhysicsComponent>(ent, out var phys) && phys.BodyType != BodyType.Static)
+                _throwing.TryThrow(ent, _random.NextAngle().ToWorldVec());
+
+            //flicker lights
+            if (_random.Prob(component.DefileEffectChance) && HasComp<PoweredLightComponent>(ent))
+                RaiseLocalEvent(ent, new GhostBooEvent());
```

Comment:
You're doing prob 3 times in a row when you should just do it once.

Also see above on GhostBooEvent. If we ever needed to make it broadcast for example you'd have to update every single area it's raised rather than the 1 method.